### ISSUE TITLE
fromTOML not working

### ISSUE THREAD

**Issue by miuirussia** [NixOS/nix#5833]
Reproduce:
change dir to any rust project, I use `git@github.com:yaxitech/ragenix.git` for example
```
$ nix run github:NixOS/nix/af553b20902b8b8efbccab5f880879b09e95eb32 repl
Welcome to Nix 2.6.0pre20211223_af553b2. Type :? for help.

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package 
error: attribute 'package' missing

       at «string»:1:1:

            1| (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
             | ^
            2|
```

Expected result:
```
nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
{ authors = [ ... ]; description = "A rust drop-in replacement for agenix"; edition = "2021"; name = "ragenix"; version = "0.1.0"; }
```

**Comment by regnat** [comment]
I guess that’s linked to https://github.com/NixOS/nix/pull/5795

/cc @nmattia

**Comment by nmattia** [comment]
I can't reproduce. This is what I did:

``` bash
[nix-shell:~/nix]$ curl -L https://raw.githubusercontent.com/yaxitech/ragenix/main/Cargo.toml -o Cargo.toml
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   805  100   805    0     0   2775      0 --:--:-- --:--:-- --:--:--  2775

[nix-shell:~/nix]$ cat Cargo.toml
[package]
name = "ragenix"
description = "A rust drop-in replacement for agenix"
version = "0.1.0"
authors = ["Vincent Haupert <vincent@yaxi.tech>"]
edition = "2021"
...

[nix-shell:~/nix]$ make clean && make install
 ...

[nix-shell:~/nix]$ ./outputs/out/bin/nix repl
Welcome to Nix 2.6.0. Type :? for help.

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
{ authors = [ ... ]; description = "A rust drop-in replacement for agenix"; edition = "2021"; name = "ragenix"; version = "0.1.0"; }

[nix-shell:~/nix]$ git rev-parse HEAD
af553b20902b8b8efbccab5f880879b09e95eb32
```

What system are you on? I'm on macOS 12.0.1, Intel

**Comment by miuirussia** [comment]
I'm on macOS 11.6.2 (20G314), you can try to use this installer: https://github.com/miuirussia/nix-unstable-installer, which gets artefacts from https://hydra.nixos.org/eval/1733072

**Comment by lilyinstarlight** [comment]
I can also reproduce this error using revision af553b2 on NixOS x86_64 and in GitHub Actions (Ubuntu 20.04 x86_64), both with `github:NixOS/nixpkgs/nixos-unstable` for the `nixpkgs` flake input

Interestingly, it does seem that `fetchTOML` always returns a set with all of the attributes, but accessing certain ones randomly fail and which attributes fail seems to change with each new `nix repl` session

Local machine (NixOS, x86_64):
```
[I] lily@bina ~/s/nix (master)> git rev-parse HEAD
af553b20902b8b8efbccab5f880879b09e95eb32

[I] lily@bina ~/s/nix (master)> nix shell

[I] lily@bina ~/s/nix (master)> curl -L https://raw.githubusercontent.com/yaxitech/ragenix/main/Cargo.toml -o Cargo.toml
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   946  100   946    0     0   2649      0 --:--:-- --:--:-- --:--:--  2642

[I] lily@bina ~/s/nix (master)> cat Cargo.toml
[package]
name = "ragenix"
description = "A rust drop-in replacement for agenix"
version = "0.1.0"
authors = ["Vincent Haupert <vincent@yaxi.tech>"]
edition = "2021"
...

[I] lily@bina ~/s/nix (master)> nix repl
Welcome to Nix 2.6.0pre20211223_af553b2. Type :? for help.

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
error: attribute 'package' missing

       at «string»:1:1:

            1| (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
             | ^
            2|

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml))
{ build-dependencies = { ... }; dependencies = { ... }; dev-dependencies = { ... }; features = { ... }; package = { ... }; }

[I] lily@bina ~/s/nix (master)> nix repl
Welcome to Nix 2.6.0pre20211223_af553b2. Type :? for help.

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml)).package
{ authors = [ ... ]; description = "A rust drop-in replacement for agenix"; edition = "2021"; name = "ragenix"; version = "0.1.0"; }

nix-repl> (builtins.fromTOML (builtins.readFile ./Cargo.toml))
{ build-dependencies = { ... }; dependencies = { ... }; dev-dependencies = { ... }; features = { ... }; package = { ... }; }
```

GitHub Actions (Ubuntu 20.04, x86_64): https://github.com/lilyinstarlight/foosteros/runs/4633945013?check_suite_focus=true#step:6:33 (poetry2nix uses `fetchTOML` for `poetryLock`)

**Comment by edolstra** [comment]
Repro:

```
$ nix build --dry-run github:edolstra/sst
error: attribute 'package' missing

       at /nix/store/sxi9i8v1ylx8gn4lpa6nrcbncia4ldqz-source/flake.nix:68:58:

           67|
           68|           (builtins.filter (pkg: pkg.source or "" != "") lockFile'.package);
             |                                                          ^
           69|
(use '--show-trace' to show detailed location information)
```

**Comment by lilyinstarlight** [comment]
I am very unfamiliar with the Nix codebase, but I noticed one functional difference in 10a8b5d with turning TOML tables into attrsets:

Do attrs in new sets need to be sorted? It looks like a call to `v.attrs->sort()` was in the cpptoml-based code, but not in the new toml11-based code ([here in the diff](https://github.com/NixOS/nix/commit/10a8b5d3ae65cbf842faa83a91e7805276cb799e#diff-1d978fb95572cff6de29d841f96cee8be698acc26ebcaefb2b9d756e786bf881L40))

Adding `v.attrs->sort();` between lines 33 and 34 in fromTOML.cc ([here](https://github.com/NixOS/nix/blob/0e90b13ab1df925e549b5d55853b65911b4b40d3/src/libexpr/primops/fromTOML.cc#L33-L34)) seemed to fix the issue on my local machine, but I don't know enough about the internals of Nix to know whether that's a proper fix or not (or if it is only working now locally by chance)

**Comment by balsoft** [comment]
@lilyinstarlight as far as I can tell (my Nix codebase knowledge is also limited), your fix is correct. Would you mind opening a pull request with this change?

**Comment by lilyinstarlight** [comment]
@balsoft Done! The PR for the fix is #5841

**Comment by nmattia** [comment]
Just catching up now. Thanks a lot @lilyinstarlight, apologies for missing this in #5795 !
